src/System.Management.Automation/engine/pipeline.cs

```diff
@@ -470,295 +466,384 @@ internal Array SynchronousExecuteEnumerate(object input)
-                    toRethrowInfo = ExceptionDispatchInfo.Capture(rte);
+                    toRethrowInfo = GetFirstError(e);
                 }
 
-                this.LogExecutionException(toRethrowInfo.SourceException);
+                // By rethrowing the exception outside of the handler, we allow the CLR on X64/IA64 to free from the stack
+                // the exception records related to this exception.
+
+                // The only reason we should get here is an exception should be rethrown.
+                Diagnostics.Assert(toRethrowInfo != null, "Alternate protocol path failure");
+                toRethrowInfo.Throw();
+
+                // UNREACHABLE
+                return null;
             }
             finally
             {
+                Clean();
                 DisposeCommands();
             }
-
-            // By rethrowing the exception outside of the handler,
-            // we allow the CLR on X64/IA64 to free from the stack
-            // the exception records related to this exception.
-
-            // The only reason we should get here is if
-            // an exception should be rethrown.
-            Diagnostics.Assert(toRethrowInfo != null, "Alternate protocol path failure");
-            toRethrowInfo.Throw();
-            return null; // UNREACHABLE
         }
 
         /// <summary>
-        /// This routine starts the stepping process. It is optional to
-        /// call this but can be useful if you want the begin clauses
-        /// of the pipeline to be run even when there may not be any input
-        /// to process as is the case for I/O redirection into a file. We
-        /// still want the file opened, even if there was nothing to write to it.
+        /// This routine starts the stepping process. It is optional to call this but can be useful
+        /// if you want the begin clauses of the pipeline to be run even when there may not be any
+        /// input to process as is the case for I/O redirection into a file. We still want the file
+        /// opened, even if there was nothing to write to it.
         /// </summary>
         /// <param name="expectInput">True if you want to write to this pipeline.</param>
         internal void StartStepping(bool expectInput)
         {
+            bool startSucceeded = false;
             try
             {
                 Start(expectInput);
+                startSucceeded = true;
 
-                // If a terminating error occurred, report it now.
-                if (_firstTerminatingError != null)
-                {
-                    _firstTerminatingError.Throw();
-                }
+                // Check if this pipeline is being stopped asynchronously.
+                ThrowFirstErrorIfExisting(logException: false);
             }
-            catch (PipelineStoppedException)
+            catch (Exception e)
             {
+                Clean();
                 DisposeCommands();
```

Comment:
Wow! Did we have a bug with leaking resources?

Reply:
Yes, it looks like so. The `internal Array Step(object input)` method below this one always has the `catch (Exception)` block where `DisposeCommands()` is called.